build.rs

```diff
@@ -1015,3 +1015,9 @@ fn prefix_all_symbols(pp: char, prefix_prefix: &str, prefix: &str) -> String {
 
     out
 }
+
+// Avoid assuming the working directory is the same is the $CARGO_MANIFEST_DIR so that toolchains
+// which may assume other working directories can still build this code.
+fn absolute_path_from_manifest_dir_relative_path<P: AsRef<Path>>(path: P) -> PathBuf {
```

Comment:
The docs for `Path::join` say that it handles absolute paths fine so I suggest we drop the "_from_relative_path" part of the name.

I suggest that we avoid having low-level functions in build.rs call this function, because it makes all the logic dependent on everything being compatible with Cargo w.r.t. the environment variables set. Instead, I suggest that at the top level, we compute `c_root_dir: PathBuf` and thread that through the functions that need it; those functions would then `join` paths to `c_root_dir` as necessary. We need to move to a design that is more like this anyway as we're working on generators for BUILD.gn and Bazel BUILD files that reuse some of this logic; they likely won't read any of these environment variables.